### PR TITLE
Add ZXing-C++ QR scanner

### PR DESCRIPTION
**Pull request by EchterAgo**
This is a test for using [ZXing-C++](https://github.com/zxing-cpp/zxing-cpp) as a QR scanner. The zbar scanner sometimes can not decode codes for some reason, the ZXing scanner seems to be much more reliable.

To test this you need to install the [zxing-cpp](https://pypi.org/project/zxing-cpp/) package from PyPi. On Windows systems you need to install version 2.0.0 because there is an issue with the MSVC build, see https://github.com/zxing-cpp/zxing-cpp/issues/612 and https://github.com/zxing-cpp/zxing-cpp/pull/613

The next release of `zxing-cpp` will have the MSVC build issue fixed and use the buffer protocol for data sharing so numpy is not required anymore.

TODO:

- [x] Set the parameters for the scanner better, we should only search QR codes and plain text.
- [x] Fix ZXing-cpp Python wrapper compilation with MSVC.
- [x] Build system integration.
- [x] Should we change the default or make it configurable?
- [x] This currently requires `numpy`, should we modify the bindings?

### CODE REVIEW

**Comment by EchterAgo** [comment]
Updated to only search for QR codes and plain text.

**Comment by EchterAgo** [comment]
This currently requires numpy, maybe we should modify the Python bindings to also take ctypes pointers and make ctypes optional.

**Comment by EchterAgo** [comment]
There is also discussion about zxing-cpp Python bindings without numpy here: https://github.com/zxing-cpp/zxing-cpp/issues/283

**Comment by EchterAgo** [comment]
This is what I needed to get zxing to work with Python buffers, which Numpy arrays are compatible with: https://github.com/EchterAgo/zxing-cpp/commit/fb45cef3b9c8f87598fd96044c2e93bf368e9696

**Comment by fyookball** [comment]
Looks reasonable at a glance.  Seems like most of the actual code for the scanner is pulled from `import zxingcpp`.

**Comment by EchterAgo** [comment]
I managed to fix the MSVC compilation issue in the zxing-cpp Python wrapper, see https://github.com/zxing-cpp/zxing-cpp/pull/613

**Comment by cculianu** [comment]
> I managed to fix the MSVC compilation issue in the zxing-cpp Python wrapper, see [zxing-cpp/zxing-cpp#613](https://github.com/zxing-cpp/zxing-cpp/pull/613)

Oh wow that's cool man. Nice! 👏

**Comment by EchterAgo** [comment]
Updated to not require numpy, depends on https://github.com/zxing-cpp/zxing-cpp/pull/615 which still needs some adjustments.

**Comment by EchterAgo** [comment]
It looks like removing `numpy` from `zxing-cpp` is doable but complicated, I still need to update and clean up the PR for that a bit, but I think we will have this soon.

As an alternative, we could use our own fork of `zxing-cpp` as a git submodule and install it on build. This would not be ideal for some Linux distros, but they could also just add a `numpy` dependency.

**Comment by EchterAgo** [comment]
https://github.com/zxing-cpp/zxing-cpp/pull/615 was merged

**Comment by cculianu** [comment]
Is this ready? Did they do a release yet?

**Comment by EchterAgo** [comment]
Yes, they did a release, I am working on getting it building now.

**Comment by EchterAgo** [comment]
This should be ready. I tested on Windows and Linux but I don't see why it wouldn't work on macOS.

Note that I've added a QR reader selection which on macOS also lists the native system QR reader, so we might want to check that it still works.

**Comment by EchterAgo** [comment]
This also includes / depdends on #2752 and #2753

**Comment by cculianu** [comment]
I should probably try this against macOS since that's the only place it's untested.. right?

**Comment by cculianu** [comment]
- Reading QR from screen works now on macOS (was broken before). Yay!
- Reading QT from camera works better now (was flakier before). Yay!
- HOWEVER: If I try to read QR from png file via the Send dialog's little "file" icon in the "Pay to" text edit, I get this stack trace now:

```
Traceback (most recent call last):
  File "/Users/calin/src/ec/electron-cash/electroncash_gui/qt/qrtextedit.py", line 74, in file_input
    scanned_qrs = self.scan_qr_from_image(image)
  File "/Users/calin/src/ec/electron-cash/electroncash_gui/qt/qrtextedit.py", line 117, in scan_qr_from_image
    res = qr_reader.read_qr_code(
  File "/Users/calin/src/ec/electron-cash/electroncash/qrreaders/zxing.py", line 59, in read_qr_code
    pybuffer = _PyMemoryView_FromMemory(buffer, buffer_size, _PyBUF_READ).cast("B", (height, width))
TypeError: memoryview: product(shape) * itemsize != buffer size
```

Maybe some bug?

File: electroncash/qrreaders/zxing.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+# BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+# ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+import ctypes
+from typing import List
+
+import zxingcpp
```
Comment (cculianu) [inline electroncash/qrreaders/zxing.py:27]: As a courtesy for users running from dev source.. might be helpful to give them this message. What do you think? I know when I first ran this I had to like scour the requirements-binaries.txt to get the version we need and so perhaps just having it printed in the output may be helpful. What do you think?

```suggestion
try:
    import zxingcpp
except ModuleNotFoundError:
    import sys
    print("\n\nzxing-cpp module missing. Please try installing it with: python3 -m pip install 'zxing-cpp>=2.2.0'"
          " --user\n\n", file=sys.stderr)
    raise
```

**Comment by cculianu** [comment]
Can we nuke the macOS QR reader? Just like nuke it as an option.. delete the .py file.. everything. 

It requires the building of some lib which is now disabled for macOS anyway. The lib is even missing entirely from the source tree and is not accessible.

Here is the stack trace if I try and use the mac QR reader on OSX:

```
the macos one requires a lib ... a lib we no longer build I believe. I would just nuke that man. zxing-cpp seems to work better anyway right now there's no reason to keep the macos one around. i will write in the PR that this should be nuked.

I'm getting this stack trace now when i try to use it:

Traceback (most recent call last):
  File "/Users/calin/src/ec/electron-cash/electroncash/qrreaders/osxqrdetect.py", line 97, in __init__
    cls.LIB = ctypes.cdll.LoadLibrary(os.path.join(lib_dir, self.LIBNAME))
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/ctypes/__init__.py", line 451, in LoadLibrary
    return self._dlltype(name)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/ctypes/__init__.py", line 373, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: dlopen(/Users/calin/src/ec/electron-cash/contrib/osx/OSXQRDetect/build/Release/libosxqrdetect.dylib, 0x0006): tried: '/Users/calin/src/ec/electron-cash/contrib/osx/OSXQRDetect/build/Release/libosxqrdetect.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/calin/src/ec/electron-cash/contrib/osx/OSXQRDetect/build/Release/libosxqrdetect.dylib' (no such file), '/Users/calin/src/ec/electron-cash/contrib/osx/OSXQRDetect/build/Release/libosxqrdetect.dylib' (no such file)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/calin/src/ec/electron-cash/electroncash_gui/qt/qrtextedit.py", line 74, in file_input
    scanned_qrs = self.scan_qr_from_image(image)
  File "/Users/calin/src/ec/electron-cash/electroncash_gui/qt/qrtextedit.py", line 110, in scan_qr_from_image
    qr_reader = get_qr_reader(self.config.get("qr_reader", None))
  File "/Users/calin/src/ec/electron-cash/electroncash/qrreaders/__init__.py", line 82, in get_qr_reader
    return reader_type()
  File "/Users/calin/src/ec/electron-cash/electroncash/qrreaders/osxqrdetect.py", line 100, in __init__
    raise MissingLib from e
electroncash.qrreaders.MissingLib
```

File: electroncash/qrreaders/zxing.py (hunk 1 above)
Comment (EchterAgo) [reply electroncash/qrreaders/zxing.py:27]: Added that as well as handling the `ModuleNotFoundError` for detecting which QR readers are available.

**Comment by EchterAgo** [comment]
> Can we nuke the macOS QR reader? Just like nuke it as an option.. delete the .py file.. everything.
> 
> It requires the building of some lib which is now disabled for macOS anyway. The lib is even missing entirely from the source tree and is not accessible.

Yea, the detection code was not right yet, it didn't call the constructor so it never got to see the `MissingLib` exception. Now it should detect that the OSX QR reader is not available and just not list it.

**Comment by EchterAgo** [comment]
> * HOWEVER: If I try to read QR from png file via the Send dialog's little "file" icon in the "Pay to" text edit, I get this stack trace now:

Can you share the png image please?

**Comment by cculianu** [comment]
> > * HOWEVER: If I try to read QR from png file via the Send dialog's little "file" icon in the "Pay to" text edit, I get this stack trace now:
> 
> Can you share the png image please?

Yes. I will do so when I get back to my desk (I’m running errands now out and about). Maybe in 1-2 h.

thanks for being responsive

**Comment by cculianu** [comment]
> > * HOWEVER: If I try to read QR from png file via the Send dialog's little "file" icon in the "Pay to" text edit, I get this stack trace now:
> 
> Can you share the png image please?

Here is the file. It was saved by EC itself in the "Receive" dialog.
qrcode

**Comment by EchterAgo** [comment]
So there seems to be padding in the file, the rows should be 450 pixel wide but it stores 452 bytes per row, so in the end `width * height` does not equal the number of bytes that was expected. This seems to be specific about this image, other images work for me. I'll see what I can do about it.

**Comment by cculianu** [comment]
> So there seems to be padding in the file, the rows should be 450 pixel wide but it stores 452 bytes per row, so in the end `width * height` does not equal the number of bytes that was expected. This seems to be specific about this image, other images work for me. I'll see what I can do about it.

Generated by EC :/.

Thanks for looking into it.

**Comment by EchterAgo** [comment]
Yea, so Qt aligns the images lines to 4 bytes, so if the width is not divisible by 4 then `bytesPerLine != width`, which means the memoryview is non contigous. This is supported by setting the `strides` of the `memoryview`, but I haven't been able to figure out how to set it after `PyMemoryView_FromMemory`. We might have to flatten the buffer.

I'm currently swapping my server into a new case, after that I'll try to fix it.

**Comment by cculianu** [comment]
I couldn't help myself and I created a patch to fix this. The caller just ensures the image data is de-padded. Let me know if there is a better place to put this code or if this is ok.. or maybe there is a "faster" way to do this with PyMemoryView?

https://github.com/EchterAgo/Electron-Cash/pull/4

**Comment by EchterAgo** [comment]
FYI, here is some info on why strides are not possible with `zxing-c++` yet:

https://github.com/python/cpython/issues/85398
https://discuss.python.org/t/functionality-for-working-with-different-strides-memoryview-s/15714

It should work in `zxing-c++` if Python actually had a way to change the strides of a `memoryview` using constructor, cast or C API.

**Pull request merged**